File: IBJts_10.44.01/samples/Java/samples/testbed/config/ConfigSamples.java
/* Copyright (C) 2026 Interactive Brokers LLC. All rights reserved. This code is subject to the terms
 * and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable. */

package samples.testbed.config;

import com.ib.client.protobuf.ApiConfigProto;
import com.ib.client.protobuf.ApiSettingsConfigProto;
import com.ib.client.protobuf.MessageConfigProto;
import com.ib.client.protobuf.OrdersConfigProto;
import com.ib.client.protobuf.OrdersSmartRoutingConfigProto;
import com.ib.client.protobuf.UpdateConfigRequestProto;
import com.ib.client.protobuf.UpdateConfigWarningProto;

public class ConfigSamples {

    public static UpdateConfigRequestProto.UpdateConfigRequest UpdateConfigApiSettings(int reqId) {
        //! [UpdateApiSettingsConfig]
        UpdateConfigRequestProto.UpdateConfigRequest.Builder updateConfigRequestBuilder = UpdateConfigRequestProto.UpdateConfigRequest.newBuilder();
        ApiConfigProto.ApiConfig.Builder apiConfigBuilder = ApiConfigProto.ApiConfig.newBuilder();

        ApiSettingsConfigProto.ApiSettingsConfig.Builder apiSettingsConfigBuilder = ApiSettingsConfigProto.ApiSettingsConfig.newBuilder();
        apiSettingsConfigBuilder.setTotalQuantityForMutualFunds(true);
        apiSettingsConfigBuilder.setDownloadOpenOrdersOnConnection(true);
        apiSettingsConfigBuilder.setIncludeVirtualFxPositions(true);
        apiSettingsConfigBuilder.setPrepareDailyPnL(true);
        apiSettingsConfigBuilder.setSendStatusUpdatesForVolatilityOrders(true);
        apiSettingsConfigBuilder.setEncodeApiMessages("osCodePage");
        apiSettingsConfigBuilder.setSocketPort(7497);
        apiSettingsConfigBuilder.setUseNegativeAutoRange(true);
        apiSettingsConfigBuilder.setCreateApiMessageLogFile(true);
        apiSettingsConfigBuilder.setIncludeMarketDataInLogFile(true);
        apiSettingsConfigBuilder.setExposeTradingScheduleToApi(true);
        apiSettingsConfigBuilder.setSplitInsuredDepositFromCashBalance(true);
        apiSettingsConfigBuilder.setSendZeroPositionsForTodayOnly(true);
        apiSettingsConfigBuilder.setUseAccountGroupsWithAllocationMethods(true);
        apiSettingsConfigBuilder.setLoggingLevel("error");
        apiSettingsConfigBuilder.setMasterClientId(3);
        apiSettingsConfigBuilder.setBulkDataTimeout(25);
        apiSettingsConfigBuilder.setComponentExchSeparator("#");
        apiSettingsConfigBuilder.setRoundAccountValuesToNearestWholeNumber(true);
        apiSettingsConfigBuilder.setShowAdvancedOrderRejectInUi(true);
        apiSettingsConfigBuilder.setRejectMessagesAboveMaxRate(true);
        apiSettingsConfigBuilder.setMaintainConnectionOnIncorrectFields(true);
        apiSettingsConfigBuilder.setCompatibilityModeNasdaqStocks(true);
        apiSettingsConfigBuilder.setSendInstrumentTimezone("utc");
        apiSettingsConfigBuilder.setSendForexDataInCompatibilityMode(true);
        apiSettingsConfigBuilder.setMaintainAndResubmitOrdersOnReconnect(true);
        apiSettingsConfigBuilder.setHistoricalDataMaxSize(4);
        apiSettingsConfigBuilder.setAutoReportNettingEventContractTrades(true);
        apiSettingsConfigBuilder.setOptionExerciseRequestType("final");
        apiSettingsConfigBuilder.addTrustedIPs("127.0.0.1");
        
        apiConfigBuilder.setSettings(apiSettingsConfigBuilder.build());
        updateConfigRequestBuilder.setReqId(reqId);
        updateConfigRequestBuilder.setApi(apiConfigBuilder.build());

        return updateConfigRequestBuilder.build();
        //! [UpdateApiSettingsConfig]
    }

    public static UpdateConfigRequestProto.UpdateConfigRequest UpdateOrdersConfig(int reqId) {
        //! [UpdateOrderConfig]
        UpdateConfigRequestProto.UpdateConfigRequest.Builder updateConfigRequestBuilder = UpdateConfigRequestProto.UpdateConfigRequest.newBuilder();
        OrdersConfigProto.OrdersConfig.Builder ordersConfigBuilder = OrdersConfigProto.OrdersConfig.newBuilder();
        OrdersSmartRoutingConfigProto.OrdersSmartRoutingConfig.Builder ordersSmartRoutingConfigBuilder = OrdersSmartRoutingConfigProto.OrdersSmartRoutingConfig.newBuilder();
        ordersSmartRoutingConfigBuilder.setSeekPriceImprovement(true);
        ordersSmartRoutingConfigBuilder.setDoNotRouteToDarkPools(true);
        ordersConfigBuilder.setSmartRouting(ordersSmartRoutingConfigBuilder.build());
        updateConfigRequestBuilder.setReqId(reqId);
        updateConfigRequestBuilder.setOrders(ordersConfigBuilder.build());
        return updateConfigRequestBuilder.build();
        //! [UpdateOrderConfig]
    }

    public static UpdateConfigRequestProto.UpdateConfigRequest UpdateMessageConfigConfirmMandatoryCapPriceAccepted(int reqId) {
        //! [UpdateMessageConfigConfirmMandatoryCapPriceAccepted]
        UpdateConfigRequestProto.UpdateConfigRequest.Builder updateConfigRequestBuilder = UpdateConfigRequestProto.UpdateConfigRequest.newBuilder();

        MessageConfigProto.MessageConfig.Builder messageConfigBuilder = MessageConfigProto.MessageConfig.newBuilder();
        messageConfigBuilder.setId(131);
        messageConfigBuilder.setEnabled(false);
        updateConfigRequestBuilder.setReqId(reqId);
        updateConfigRequestBuilder.addMessages(messageConfigBuilder.build());

        UpdateConfigWarningProto.UpdateConfigWarning.Builder updateConfigWarningBuilder = UpdateConfigWarningProto.UpdateConfigWarning.newBuilder();
        updateConfigWarningBuilder.setMessageId(131);
        updateConfigRequestBuilder.addAcceptedWarnings(updateConfigWarningBuilder.build());

        return updateConfigRequestBuilder.build();
        //! [UpdateMessageConfigConfirmMandatoryCapPriceAccepted]
    }

    public static UpdateConfigRequestProto.UpdateConfigRequest UpdateConfigOrderIdReset(int reqId) {
        //! [ UpdateConfigOrderIdReset]
        UpdateConfigRequestProto.UpdateConfigRequest.Builder updateConfigRequestBuilder = UpdateConfigRequestProto.UpdateConfigRequest.newBuilder();
        updateConfigRequestBuilder.setReqId(reqId);
        updateConfigRequestBuilder.setResetAPIOrderSequence(true);
        return updateConfigRequestBuilder.build();
        //! [ UpdateConfigOrderIdReset]
    }

}

File: IBJts_10.44.01/samples/Cpp/TestCppClient/ConfigSamples.cpp
/* Copyright (C) 2026 Interactive Brokers LLC. All rights reserved. This code is subject to the terms
 * and conditions of the IB API Non-Commercial License or the IB API Commercial License, as applicable. */
#include "StdAfx.h"

#include "ConfigSamples.h"

protobuf::UpdateConfigRequest ConfigSamples::UpdateConfigApiSettings(int reqId)
{
    //! [UpdateApiSettingsConfig]
    protobuf::UpdateConfigRequest updateConfigRequestProto;
    protobuf::ApiConfig apiConfigProto;
    protobuf::ApiSettingsConfig apiSettingsConfigProto;

    apiSettingsConfigProto.set_totalquantityformutualfunds(true);
    apiSettingsConfigProto.set_downloadopenordersonconnection(true);
    apiSettingsConfigProto.set_includevirtualfxpositions(true);
    apiSettingsConfigProto.set_preparedailypnl(true);
    apiSettingsConfigProto.set_sendstatusupdatesforvolatilityorders(true);
    apiSettingsConfigProto.set_encodeapimessages("osCodePage");
    apiSettingsConfigProto.set_socketport(7497);
    apiSettingsConfigProto.set_usenegativeautorange(true);
    apiSettingsConfigProto.set_createapimessagelogfile(true);
    apiSettingsConfigProto.set_includemarketdatainlogfile(true);
    apiSettingsConfigProto.set_exposetradingscheduletoapi(true);
    apiSettingsConfigProto.set_splitinsureddepositfromcashbalance(true);
    apiSettingsConfigProto.set_sendzeropositionsfortodayonly(true);
    apiSettingsConfigProto.set_useaccountgroupswithallocationmethods(true);
    apiSettingsConfigProto.set_logginglevel("error");
    apiSettingsConfigProto.set_masterclientid(3);
    apiSettingsConfigProto.set_bulkdatatimeout(25);
    apiSettingsConfigProto.set_componentexchseparator("#");
    apiSettingsConfigProto.set_roundaccountvaluestonearestwholenumber(true);
    apiSettingsConfigProto.set_showadvancedorderrejectinui(true);
    apiSettingsConfigProto.set_rejectmessagesabovemaxrate(true);
    apiSettingsConfigProto.set_maintainconnectiononincorrectfields(true);
    apiSettingsConfigProto.set_compatibilitymodenasdaqstocks(true);
    apiSettingsConfigProto.set_sendinstrumenttimezone("utc");
    apiSettingsConfigProto.set_sendforexdataincompatibilitymode(true);
    apiSettingsConfigProto.set_maintainandresubmitordersonreconnect(true);
    apiSettingsConfigProto.set_historicaldatamaxsize(4);
    apiSettingsConfigProto.set_autoreportnettingeventcontracttrades(true);
    apiSettingsConfigProto.set_optionexerciserequesttype("final");
    apiSettingsConfigProto.add_trustedips()->append("127.0.0.1");
    apiConfigProto.mutable_settings()->CopyFrom(apiSettingsConfigProto);
    updateConfigRequestProto.set_reqid(reqId);
    updateConfigRequestProto.mutable_api()->CopyFrom(apiConfigProto);

    return updateConfigRequestProto;
    //! [UpdateApiSettingsConfig]
}

protobuf::UpdateConfigRequest ConfigSamples::UpdateOrdersConfig(int reqId)
{
    //! [UpdateOrderConfig]
    protobuf::UpdateConfigRequest updateConfigRequestProto;
    protobuf::OrdersConfig ordersConfigProto;
    protobuf::OrdersSmartRoutingConfig ordersSmartRoutingConfigProto;
    ordersSmartRoutingConfigProto.set_seekpriceimprovement(true);
    ordersSmartRoutingConfigProto.set_donotroutetodarkpools(true);
    ordersConfigProto.mutable_smartrouting()->CopyFrom(ordersSmartRoutingConfigProto);
    updateConfigRequestProto.set_reqid(reqId);
    updateConfigRequestProto.mutable_orders()->CopyFrom(ordersConfigProto);
    return updateConfigRequestProto;
    //! [UpdateOrderConfig]
}

protobuf::UpdateConfigRequest ConfigSamples::UpdateMessageConfigConfirmMandatoryCapPriceAccepted(int reqId)
{
    //! [UpdateMessageConfigConfirmMandatoryCapPriceAccepted]
    protobuf::UpdateConfigRequest updateConfigRequestProto;
    protobuf::MessageConfig messageConfigProto;
    messageConfigProto.set_id(131);
    messageConfigProto.set_enabled(false);
    updateConfigRequestProto.set_reqid(reqId);
    updateConfigRequestProto.add_messages()->CopyFrom(messageConfigProto);
    protobuf::UpdateConfigWarning updateConfigWarningProto;
    updateConfigWarningProto.set_messageid(131);
    updateConfigRequestProto.add_acceptedwarnings()->CopyFrom(updateConfigWarningProto);
    return updateConfigRequestProto;
    //! [UpdateMessageConfigConfirmMandatoryCapPriceAccepted]
}

protobuf::UpdateConfigRequest ConfigSamples::UpdateConfigOrderIdReset(int reqId)
{
    //! [ UpdateConfigOrderIdReset]
    protobuf::UpdateConfigRequest updateConfigRequestProto;
    updateConfigRequestProto.set_reqid(reqId);
    updateConfigRequestProto.set_resetapiordersequence(true);
    return updateConfigRequestProto;
    //! [ UpdateConfigOrderIdReset]
}
